src/hooks/useResizeObserver.ts

```diff
@@ -0,0 +1,18 @@
+import { RefObject, useLayoutEffect } from 'react'
+import ResizeObserver from 'resize-observer-polyfill'
+
+/** Hoke to add resizeObserver on the passed ref and call the callback function on resize. */
+const useResizeObserver = (ref: RefObject<HTMLElement>, onResize: () => void) => {
+
+  useLayoutEffect(() => {
+    if (ref.current === null) {
+      return;
+    }
+
+    const ro = new ResizeObserver(onResize);
+    ro.observe(ref.current!);
+
+    return () => ro.disconnect()
+  }, []);
+}
```

Comment:
Looks good! But we can make it tighter with these changes.

```javascript
const useResizeObserver = (ref: RefObject<HTMLElement>, onResize: (target: Element) => void) => {
  const [ro] = useState(() => new ResizeObserver(([entry]) => onResize(entry.target)))
  useLayoutEffect(() => {
    if (ref.current) ro.observe(ref.current)
    return () => ro.disconnect()
  }, []);
}
```
So we don't have do access ref again in the callback.

```javascript
useResizeObserver(toolbarRef, target => {
    const { scrollWidth, clientWidth } = target
    updateToolbarOrigin(scrollWidth - clientWidth > TOOLBAR_WRAPPER_DELTA ? toolbarOrigins.topRight : toolbarOrigins.topLeft)
})
```

Reply:
> Looks good! But we can make it tighter with these changes.
> 
> ```js
> const useResizeObserver = (ref: RefObject<HTMLElement>, onResize: (target: Element) => void) => {
>   const [ro] = useState(() => new ResizeObserver(([entry]) => onResize(entry.target)))
>   useLayoutEffect(() => {
>     if (ref.current) ro.observe(ref.current)
>     return () => ro.disconnect()
>   }, []);
> }
> ```
> 
> So we don't have do access ref again in the callback.

With regard to splitting the logic between `useState` and `useLayoutEffect`, I'm not sure quite sure about that because we're creating `ResizeObserver` instance without even checking if the current reference has a valid value, thus making it a dead store in some cases.
`if (ref.current), then initialize and observer`
Ideally, we should get rid of the null-check altogether using TS but that'd be possible when `Toolbar.js` gets migrated to ts (hopefully soon). But for now, we'll have to live with it.
What do you think?

> 
> ```js
> useResizeObserver(toolbarRef, target => {
>     const { scrollWidth, clientWidth } = target
>     updateToolbarOrigin(scrollWidth - clientWidth > TOOLBAR_WRAPPER_DELTA ? toolbarOrigins.topRight : toolbarOrigins.topLeft)
> })
> ```

Passing the target element in `onResize` callback definitely helps with some logic abstraction and I'll update that.